generate-release-cycle.py

```diff
@@ -0,0 +1,136 @@
+"""
+Read in a JSON and generate two CSVs and a Mermaid file
+"""
+from __future__ import annotations
+
+import csv
+import datetime as dt
+import json
+
+MERMAID_HEADER = """
+gantt
+    dateFormat  YYYY-MM-DD
+    title       Python release cycle
+    axisFormat  %Y
+""".lstrip()
+
+MERMAID_SECTION = """
+    section Python {version}
+    {release_status}               :{mermaid_status}   python{version}, {release_date},{eol}
+"""  # noqa: E501
+
+MERMAID_STATUS_MAPPING = {
+    "feature": "",
+    "bugfix": "active,",
+    "security": "done,",
+    "end-of-life": "crit,",
+}
+
+CSV_HEADER = (
+    "Branch",
+    "Schedule",
+    "Status",
+    "First release",
+    "End of life",
+    "Release manager",
+)
+
+
+def pep(number: int) -> str:
+    """Format PEP number with Sphinx role"""
+    return f":pep:`{number}`"
+
+
+def csv_date(date_str: str, now_str: str) -> str:
+    """Format a date for CSV"""
+    if date_str > now_str:
+        # Future, add italics
+        return f"*{date_str}*"
+    return date_str
+
+
+def mermaid_date(date_str: str) -> str:
+    """Format a date for Mermaid"""
+    if len(date_str) == len("yyyy-mm"):
+        # Mermaid needs a full yyyy-mm-dd, so let's approximate
+        date_str = f"{date_str}-01"
+    return date_str
+
+
+class Versions:
+    """For converting JSON to CSV and Mermaid"""
+
+    def __init__(self) -> None:
+        with open("include/release-cycle.json") as f:
+            self.versions = json.load(f)
+
+    def save_csv(self) -> None:
+        """Output CSV files"""
+        now_str = str(dt.datetime.utcnow())
+
+        with (
+            open(
+                "include/branches.csv", "w", encoding="UTF-8", newline=""
+            ) as file_branches,
+            open(
+                "include/end-of-life.csv", "w", encoding="UTF-8", newline=""
+            ) as file_eol,
+        ):
+            csv_branches = csv.writer(
+                file_branches, quoting=csv.QUOTE_MINIMAL, lineterminator="\n"
+            )
+            csv_eol = csv.writer(
+                file_eol, quoting=csv.QUOTE_MINIMAL, lineterminator="\n"
+            )
+
+            csv_branches.writerow(CSV_HEADER)
+            csv_eol.writerow(CSV_HEADER)
+
+            sorted_versions = sorted(
+                self.versions.items(),
+                key=lambda k: int(k[0].replace(".", "")),
+                reverse=True,
+            )
```

Comment:
This won't actually work quite right, as it will sort all `X.YY` versions ahead of all `X.Y` versions. My refactored version, https://github.com/CAM-Gerlach/devguide/commit/c2cbf6b0326d6b63c3a35934529c8247f83d5785 , replaces the `key` line with `lambda k: [int(i) for i in k.split(".")]` which performs the same sorting as with standard version tuples.

Reply:
> That's the intention:
> 
> ```python
> >>> sorted(["30", "31", "310", "311"], reverse=True)
> ['311', '310', '31', '30']
> ```

I'm a little confused what you mean here—the sorting algorithm being used here is sorting these as integers, not as strings. In any case, this algorithm does not give the correct result for any two-digit minor versions 2 or higher, which are not included in the above set:

```python
>>> sorted(["30", "31", "32", "310", "311"], reverse=True)
['32', '311', '310', '31', '30']
```

whereas tuple sorting sorts this correctly:

```python
>>> sorted([(3, 0), (3, 1), (3, 2), (3, 10), (3, 11)], reverse=True)
[(3, 11), (3, 10), (3, 2), (3, 1), (3, 0)]
```

Moreover, the actual sorting here is done by casting the above strings to integers, and while that does happen to work on the currently-released Python versions, it only does so by accident because there happens to be only one major version that has two-digit minor versions, and _that_ in turn happens to be the latest version. Look what happens when you add a `4.0`:

```python
>>> sorted([30, 31, 310, 311, 40], reverse=True)
[311, 310, 40, 31, 30]
```

Whereas again, tuple sorting handles this correctly:

```python
>>> sorted([(3, 0), (3, 1), (3, 2), (3, 10), (3, 11), (4, 0)], reverse=True)
[(4, 0), (3, 11), (3, 10), (3, 2), (3, 1), (3, 0)]
```

Reply:
Ah yes, I had missed the 4.0 case, thanks!

3.x->4.0 will be a good source of bugs just for version handling, https://github.com/asottile/flake8-2020 can help with some. And a common fix is to compare tuples not strings :)

Reply:
Maybe you're right and we do need that `flake8-2020` in the pre-commit config after all :laughing: